[wip] Add llvm 21 support

mcbarton opened the pull request:
# Description

Please include a summary of changes, motivation and context for this PR.

Fixes # (issue)

## Type of change

Please tick all options which are relevant.

- [ ] Bug fix
- [x] New feature
- [ ] Requires documentation updates

## Testing

Please describe the test(s) that you added and ran to verify your changes.

## Checklist

- [x] I have read the contribution guide recently

mcbarton commented:
> Hey @mcbarton ,
> 
> Are you working on this ?
> 
> I think I saw quite some errors errors while trying to build cppinterop against latest llvm 21 few days back.

@anutosh491 With the exception of the Windows native build you can build CppInterOp against llvm 21 with the changes in this PR (Windows fails building llvm). The tests do not pass however. Given the changes in this PR are minimal, then the source of these errors are likely on the llvm side. I will update this PR soon to get rid of the Windows Emscripten cross compile patch since it is no longer needed, and to update the documentation. Not sure I have any spare time at the moment though to work out the source of the failing tests.

Vipul-Cariappa reviewed (commented):
LGTM. Will approve if the CI is green.
@mcbarton, does this also test cppyy with LLVM 21?
@vgvassilev, could you take a look at the `.patch` files in this PR?

mcbarton commented:
Once we have the ci green, we probably need to remove a few ci jobs, since we are hitting the maximum cache size again. My suggestion is to remove the Ubuntu x86 llvm 18 and osx x86 llvm 18 jobs. This will scrap us just under the maximum I think, and we can continue to support these from the cppyy repos.

mcbarton commented:
I have opened up an issue in llvm about the failing Windows build for llvm 21 here https://github.com/llvm/llvm-project/issues/158716

anutosh491 commented:
hey @mcbarton here's the updated pr on llvm for the patch : https://github.com/llvm/llvm-project/pull/132670

Hope this helps if we need to keep it for some more time on llvm 21

mcbarton commented:
> @mcbarton can you rebase and fix the size threshold? I’d like to get that in and make a release. We should fix the emscripten size in a separate effort to avoid blocking the support of llvm21.

@vgvassilev I have rebased the PR, and removed the check for the size of the Emscripten shared library size, which I fix after release. The only thing that should be left to get the ci to go green would be to fix the native osx jobs which fail with the following error https://github.com/compiler-research/CppInterOp/actions/runs/19036208336/job/54361154631?pr=672#step:12:578 . I tried fixing it about 2 weeks ago without success. I have enabled it so this PR can be edited by maintainers. If anyone knows how to fix, then they can feel free to commit the fix to the PR.

anutosh491 commented:
> would you like me to work out a patch for llvm 21 which allows CppInterOp macos jobs to pass, or try and modify CppInterOp to work with this llvm commit which currently breaks the macos jobs? Not sure if the commit description highlights we can work with this change to llvm.

Wait I'm confused. So it is a cppinterop thing and not a llvm thing ? 

As in were we doing something wrong and we need to start maintaining a patch or is that patch at fault and we need to fix it on llvm's side ?

That looks like a 300 line odd patch and not sure anyone of us knows what it does exactly, so do we really want to maintain it ? Let's fix it the organic way if we can, can we ?

anutosh491 commented:
Why not post our concerns on that PR and address the root cause ?

anutosh491 commented:
Hmm, I am not sure we should maintain random patches which we don't understand or wasn't introduced by one of us :\

I can give it a look, once I back from vacations (Jan 2026)

mcbarton commented:
Will fix the failing Emscripten jobs after the Christmas break. CppInterOpTest/InProcessJIT.InterpreterTestEmscriptenExceptionHandling passes for llvm 20, but not for llvm 21 when building with wasm exceptions. CppInterOpTest/InProcessJIT.VariableReflectionTestVariableOffsetsWithInheritance fails/crashes for llvm 20, but not for llvm 21 where is passes when building with wasm exceptions.

mcbarton commented:
> Hey @mcbarton,
> 
> i just went through the commit you had pointed out ([llvm/llvm-project#138234](https://github.com/llvm/llvm-project/pull/138234)) I don't think the commit needs to be reverted, rather we need to adapt as per that commit.
> 
> We just need this introduce this diff to get the test passing on macos
> 
> ```
> diff --git a/lib/CppInterOp/Paths.cpp b/lib/CppInterOp/Paths.cpp
> index f3980c8..bd9099d 100644
> --- a/lib/CppInterOp/Paths.cpp
> +++ b/lib/CppInterOp/Paths.cpp
> @@ -158,7 +158,13 @@ void CopyIncludePaths(const clang::HeaderSearchOptions& Opts,
>    /// User specified include entries.
>    for (unsigned i = 0, e = Opts.UserEntries.size(); i != e; ++i) {
>      const HeaderSearchOptions::Entry& E = Opts.UserEntries[i];
> -    if (E.IsFramework && E.Group != frontend::Angled)
> +    if (E.IsFramework && E.Group != frontend::Angled && E.Group != frontend::System)
> ```
> 
> What happened is clang’s Darwin driver change now passes default framework search paths via -iframework/-internal-iframework, which get stored in HeaderSearchOptions::UserEntries as IsFramework=true with Group=frontend::System (not Angled). cppinterop assumed ‘framework implies Angled’ and aborted.
> 
> So if I add such logs such above
> 
> ```
>     if (E.IsFramework && E.Group != frontend::Angled) {
>       llvm::errs() << "[cppinterop] framework include entry with non-Angled group: "
>                   << "group=" << int(E.Group)
>                   << " path=" << E.Path << "\n";
>     }
> ```
> 
> I would see, this at the tests
> 
> ```
> ......
> 1: [ RUN      ] CppInterOpTest/InProcessJIT.InterpreterTestIncludePaths
> 1: [cppinterop] framework include entry with non-Angled group: group=2 path=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks
> 1: [cppinterop] framework include entry with non-Angled group: group=2 path=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/SubFrameworks
> 1: [cppinterop] framework include entry with non-Angled group: group=2 path=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/Library/Frameworks
> .......
> ```
> 
> This proves we must allow IsFramework entries in the System group (and ideally treat them as -iframework when emitting flags)
> 
> Try it out and hopefully we can remove the patch you had introduced reverting the llvm commit ! Tests pass locally on my maocs after the 1 line diff.

Hi @anutosh491

I have checked 'Allow edits and access to secrets by maintainers' on this PR, so if you want to apply your fix directly to the PR then you can. 

After that we should only have to get the Emscripten workflow to pass. For fixing the Emscripten workflow I don't mind disabling CppInterOpTest/InProcessJIT.VariableReflectionTestVariableOffsetsWithInheritance for llvm 20 when built with wasm exceptions, but we need to fix CppInterOpTest/InProcessJIT.InterpreterTestEmscriptenExceptionHandling for llvm 21 when built with wasm exceptions.

mcbarton commented on CMakeLists.txt:
```diff
@@ -12,6 +12,9 @@ set(CMAKE_CXX_EXTENSIONS NO)
 option(CPPINTEROP_ENABLE_TESTING "Enable the CppInterOp testing infrastructure." ON)
+if(EMSCRIPTEN)
+  option(CPPINTEROP_ENABLE_WASM_EXCEPTIONS "Link using wasm exceptions flag" OFF)
+endif()
```
Although quant stack has decided to make the change to building the stack in Emscripten forge with wasm exceptions doesn't mean it needs to force that on others. The build instructions in this repo are very clear about how to build with and without wasm exceptions (and that you need wasm exceptions if you plan to build xeus-cpp), so I don't plan on dropping this option from the PR. The idea that its superior is a matter of opinion.

Putting all this discussion on the option to one side, and changing the test as you suggest (the way it is in the PR already), and building everything with none of my changes in this PR relating to wasm exceptions and just ` EMCC_CFLAGS="-sSUPPORT_LONGJMP=wasm -fwasm-exceptions"` (like emscripten forge does it https://github.com/emscripten-forge/recipes/blob/35f77d9fb19fa5f536e707b3713742f331cf3931/recipes/recipes/emscripten_emscripten-wasm32/activate.sh#L54) will not get the test pass the llvm 21 (I tried this before Christmas).

anutosh491 reviewed (approved):
Ahh okay, this doesn't have the emscripten changes anymore ! Should be good then !
